assert: add partialDeepStrictEqual

puskin94 opened the pull request:
Fixes: https://github.com/nodejs/node/issues/50399

Took heavy inspiration from https://github.com/nodejs/node/pull/53415 , trying to push it to the finish line 🚀 

On top of it, I took the liberty of:

1. refactor the code a bit
2. wrote more documentation
3. added more tests and restructured old ones
4. implement previously reported comments / suggestions

Co-Authored-By: Cristian Barlutiu

jasnell commented on doc/api/assert.md:
```diff
@@ -2548,6 +2548,237 @@ assert.throws(throwingFirst, /Second$/);
+import assert from 'node:assert';
+
+assert.deepMatch({ a: 1, b: '2' }, { a: 1, b: 2 });
+// OK
```
How is this different from `deepStrictEqual(...)` and `deepEqual(...)`? The docs should expand on that more. As it is, these examples do not really provide enough detail to determine the correct one to choose.

puskin94 replied:
thanks for the update! I adjusted the docs now

RedYetiDev commented:
This PR adds new functionality, hence semver-minor

This PR's new functionality (IMO) is notable-change

puskin94 commented on test/parallel/test-assert-objects.js:
@RedYetiDev looking at the end result, I am not sure it is cleaner or not, but I did it anyway 😄

jasnell commented on doc/api/assert.md:
```diff
@@ -2548,6 +2548,241 @@ assert.throws(throwingFirst, /Second$/);
 Due to the confusing error-prone notation, avoid a string as the second
 argument.
 
+## `assert.deepMatch(actual, expected[, message])`
```
The naming here might be confusing given that `assert.match(...)` exists and uses regexes but this `deepMatch()` has nothing to do with regexes.

It's a bit more verbose both something like `assert.partialDeepEqual()` and `assert.partialDeepStrictEqual()` might be better / less confusing?

puskin94 replied:
@jasnell I just followed the most "approved" comment in the original PR and implemented it :) 

https://github.com/nodejs/node/issues/50399#issuecomment-2296531960

Renegade334 replied:
FWIW, +1 regarding the naming semantics. My only alternative suggestion would be `deepEqualSubset()`, which is potentially a bit more intuitive than "partial" (although ES2025 will introduce the Set-specific notion of subset matching, which could itself be a source of confusion).

puskin94 replied:
@Renegade334 to be completely honest I like the "partial" more, I have the feeling it is a little more immediate to understand what the method does.  
But, I would like to have a final decision here 😄 is there any way we can tag people to take a final decision here?

jasnell commented on doc/api/assert.md:
````diff
@@ -2548,6 +2548,241 @@ assert.throws(throwingFirst, /Second$/);
+// AssertionError
+```
+
+## `assert.includes(actual, expected[, message])`
````
I'm less convinced that this one is useful. Why not just simply use `assert(actual.includes(expected))`

BridgeAR replied:
I agree with @jasnell and don't think this should be implemented. Let us concentrate on the partial inclusion. If anyone would ever ask for something else, we can still implement more.

pmarchini commented on lib/assert.js:
```diff
@@ -608,6 +620,218 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+    return StringPrototypeIncludes(actual, expected);
+  }
+
+  return expected.every((item) => actual.some((actualItem) => {
+    // eslint-disable-next-line eqeqeq
+    return loose ? actualItem == item : actualItem === item;
+  }));
```
I think we should use primordials : `ArrayPrototypeEvery` and `ArrayPrototypeSome`

puskin94 replied:
@pmarchini I thought the same but then I looked for the non primordials usage across the codebase and, considering the huge usage of them, I was like "maybe I should not" . Converted tho 😄

pmarchini commented on lib/assert.js:
```diff
@@ -608,6 +620,218 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+      return false;
+    }
+    const safeIterator = FunctionPrototypeCall(SafeMap.prototype[SymbolIterator], actual);
+    for (const { 0: key, 1: val } of safeIterator) {
```
Should we avoid `for ... of` (https://github.com/nodejs/node/blob/b0443690f89816ca3b9b8dee7e73f4527b3ff3e0/doc/contributing/primordials.md#unsafe-array-iteration)  ?

pmarchini commented on lib/assert.js:
```diff
@@ -608,6 +620,218 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+      return false;
+    }
+    const safeIterator = FunctionPrototypeCall(SafeMap.prototype[SymbolIterator], actual);
+    for (const { 0: key, 1: val } of safeIterator) {
```
Should we avoid `for ... of`? as per https://github.com/nodejs/node/blob/b0443690f89816ca3b9b8dee7e73f4527b3ff3e0/doc/contributing/primordials.md#unsafe-array-iteration

aduh95 replied:
No, as per https://github.com/nodejs/node/blob/c42d8461b061136da2b3c04bd976d2e15b81969d/doc/contributing/primordials.md#L212-L214

pmarchini commented on lib/assert.js:
```diff
@@ -608,6 +620,218 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  const keysExpected = ReflectOwnKeys(expected);
+
+  // Handle circular references
+  if (comparedObjects.has(actual)) {
```
Do we need primordials also here?

puskin94 commented:
is there any way we can push this forward? ar far as I know there is a little bit of people waiting for this and no clear consensus on the naming convention :)

jasnell commented:
I still have concerns about whether we need everything in this PR but would very much like others to weigh in. @nodejs/test_runner @nodejs/assert ... anyone have thoughts?

ljharb commented:
re naming, I definitely think the term "match" should be reserved for regexes.

if the only difference between these new methods and deepEqual etc is that they allow extra properties, then perhaps it would make more sense as an option to the existing methods rather than entirely new ones?

puskin94 commented:
@ljharb the discussion where the "option" was discarded in favor of the new API started from this comment: https://github.com/nodejs/node/issues/50399#issuecomment-1860653681

puskin94 commented:
other method names we could consider:

`assert.subsetEqual()` `assert.containsSubset()` `assert.partialEqual()`

ljharb commented:
To clarify, the current proposal takes an "actual" and "expected", and it checks that every property on `expected` is deepEqual to the same property on `actual`? Are nested objects on `expected` checked using full or partial equality?

What about if I want to assert that a property is *not* present? is there a way to represent "never"?

puskin94 commented:
good questions!

yes, this test will pass:

```js
      {
        description: 'compares two deeply nested objects with partial equality',
        actual: { a: {nested: {property: true, some: 'other'}} },
        expected: { a: {nested: {property: true}} },
      },
```

and no, there is no way to check "not inclusion" , just if an object is a subset of another

ljharb commented:
`assert.subsetDeepEqual`? (also shouldn't the default be strict, and the extra one be "Loose"?)

puskin94 commented:
that would deviate from the already present implementations, like `deepStrictEqual` and `deepEqual`

MoLow reviewed (commented):
implementation LGTM, I agree with the concerns on naming, lets change the name so this can land?

Renegade334 commented on lib/assert.js:
```diff
@@ -608,6 +623,224 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+    throw new ERR_MISSING_ARGS('actual', 'expected');
+  }
+
+  if (!assertIncludes(actual, expected, true)) {
```
I suspect this function signature is such that the types of `actual` and `expected` need validation here, otherwise this call to `assertIncludes()` is at risk of generating TypeErrors.

Renegade334 commented on lib/assert.js:
```diff
@@ -608,6 +623,224 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+    throw new ERR_MISSING_ARGS('actual', 'expected');
+  }
+
+  if (!assertIncludes(actual, expected, false)) {
```
Ditto.

Renegade334 commented on lib/assert.js:
```diff
@@ -608,6 +623,224 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+    return loose ? isDeepEqual(actual, expected) : isDeepStrictEqual(actual, expected);
+  }
+
+  // Check if actual and expected are null or not objects
+  if (actual == null || expected == null) {
+    return false;
+  }
```
I believe that null/undefined are already covered by your `isPlainObject()` logic above.

Renegade334 commented on test/parallel/test-assert-objects.js:
I think it'd be worthwhile being more stringent with validating the `expected` errors in the throw tests, at least checking whether the error is an AssertionError, an invalid args error, etc.

RedYetiDev commented:
> +1 for `partialDeepEqual`

Agreed. Also, should this have a `notPartialDeepEqual` inverse?

puskin94 commented:
Updated:

1. names: from `deepMatch` / `deepMatchStrict` to `partialDeepEqual` / `partialDeepStrictEqual`
2. tests: added a couple of cases and better syntax
3. validations: `includes` and `includesStrict` now validate `actual` and `expected` types beforehand

Renegade334 commented on lib/assert.js:
```diff
@@ -608,6 +623,233 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  }
+};
+
+function throwIfInvalidIncludesParams(actual, expected) {
```
Does this need to be wrapped in `hideStackFrames()`?

puskin94 replied:
done 😄

mcollina reviewed (approved):
lgtm

mcollina commented:
It's probably worth mentioning in the docs about `Map` and `Set` support

puskin94 commented:
Updated the docs with additional examples showing more complicated usages, including `Map`s and `Set`s

RedYetiDev commented:
Should this have the inverses also? notPartialDeepEqual and such?

puskin94 commented:
> Should this have the inverses also? notPartialDeepEqual and such?

how would that look like? something like this?

```js
assert.notPartialDeepEqual({a: 1}, {a: 1, b: 2});
// THROWS

assert.notPartialDeepEqual({a: 3}, {a: 1, b: 2});
// DOES NOT THROW
```

ayway, I would build that in a different PR, this one is already stuffed enough imho 😄

RedYetiDev commented:
Got it, everything else seems good to me

puskin94 commented:
it feels like the CI is stuck for hours now

RedYetiDev commented:
> it feels like the CI is stuck for hours now

It looks like the CI was aborted.

Renegade334 reviewed (commented):
Note that the method names will also need to be added to `lib/internal/test_runner/test.js`:

https://github.com/nodejs/node/blob/e49cf7acfbfb5ce36f49911d22f533c7056137c2/lib/internal/test_runner/test.js#L103-L120

An equivalent change will be required in `test/parallel/test-runner-assert.js`:

https://github.com/nodejs/node/blob/e49cf7acfbfb5ce36f49911d22f533c7056137c2/test/parallel/test-runner-assert.js#L7-L25

puskin94 commented:
@Renegade334 oh, I wasn't aware of it! updated 😄

simoneb commented:
> Note that the method names will also need to be added to `lib/internal/test_runner/test.js`:
> 
> https://github.com/nodejs/node/blob/e49cf7acfbfb5ce36f49911d22f533c7056137c2/lib/internal/test_runner/test.js#L103-L120
> 
> An equivalent change will be required in `test/parallel/test-runner-assert.js`:
> 
> https://github.com/nodejs/node/blob/e49cf7acfbfb5ce36f49911d22f533c7056137c2/test/parallel/test-runner-assert.js#L7-L25

Maybe we should add an automatic way to check that this happens?

RedYetiDev commented:
Certain names are intentionally excluded from that list (E.G `strict`), so adding an lint-rule validating it is easier said than done, IMO.

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +359,233 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+      return false;
+    }
+    const safeIterator = FunctionPrototypeCall(SafeMap.prototype[SymbolIterator], actual);
+    const safeIteratorArray = ArrayFrom(safeIterator);
+
+    for (let i = 0; i < safeIteratorArray.length; i++) {
+      const { 0: key, 1: val } = safeIteratorArray[i];
```
nit: we don't need to create a throw-away array here
```suggestion

    for (const { 0: key, 1: val } of safeIterator) {
```

puskin94 replied:
@aduh95 this was intentionally done to comply with https://github.com/nodejs/node/blob/b0443690f89816ca3b9b8dee7e73f4527b3ff3e0/doc/contributing/primordials.md#unsafe-array-iteration

aduh95 replied:
As I said in https://github.com/nodejs/node/pull/54630#discussion_r1766362234, this doesn't apply here, and my suggestion still stands :)

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +359,233 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+      return false;
+    }
+    const safeIterator = FunctionPrototypeCall(SafeSet.prototype[SymbolIterator], actual);
+    const safeIteratorArray = ArrayFrom(safeIterator);
+
+    for (let i = 0; i < safeIteratorArray.length; i++) {
+      if (!SetPrototypeHas(expected, safeIteratorArray[i])) {
```
same here
```suggestion

    for (const item of safeIterator) {
      if (!SetPrototypeHas(expected, item)) {
```

puskin94 replied:
@aduh95 this was intentionally done to comply with https://github.com/nodejs/node/blob/b0443690f89816ca3b9b8dee7e73f4527b3ff3e0/doc/contributing/primordials.md#unsafe-array-iteration

aduh95 replied:
I'm well aware of this document, I've written it 😃 It's not an array, it's a safe iterator, so there's no unsafe array iteration in my suggestion.

puskin94 replied:
ahahah ops, I have learned something new then! Updating it, thanks!

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +359,233 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+    return StringPrototypeIncludes(actual, expected);
+  }
+
+  return ArrayPrototypeEvery(expected, (item) => ArrayPrototypeSome(actual, (actualItem) => {
+    // eslint-disable-next-line eqeqeq
+    return loose ? actualItem == item : actualItem === item;
+  }));
```
IMO we should just be using `Array.prototype.includes`, this really should only belong in `partialDeepEqual`. Having `assert.includes` makes perfect sense, having `assert.includesStrict` less so (or it should be an alias maybe).

puskin94 replied:
I would be kinda agreeing with you if most of the methods didn't have its own `strict` variant. If we would do what you suggested, we would have `includes` which is not doing type coercion, going agains the pattern we have for pretty much the whole `assert` module

aduh95 replied:
Another path would be to remove `includes` from this PR so `partialDeepStrictEqual` can land, and we can discuss the `includes` (and possibly `includesStrict`) in another PR, wdyt?

puskin94 replied:
@aduh95 it kinda bothers me to remove functionality considering it had been implemented already 😄 But  I understand your concerns. I would ping people to give us more feedback instead tho. We could give us a couple of days?

aduh95 replied:
Sure I was suggesting that in case you'd like the `partialDeepStrictEqual` to land without waiting since it has consensus, but it's certainly fine to wait.

puskin94 replied:
what about a small input from @danielbayley ? He was the one introducing the idea of adding `includes` and `includesStrict` into the PR, getting a bunch of upvotes too

aduh95 replied:
I tried to look for that proposal, is it https://github.com/nodejs/node/issues/50399#issuecomment-2296531960? In this comment, Daniel seems to agree with me that `assert.includes` should mirror the built-in JS methods, like I'm suggesting. Or am I missing something?

puskin94 replied:
@aduh95 I might have misunderstood the previous discussion then!

1. removed `includesStrict`
2. adapted `includes` to be mirroring the native `includes`
3. adapted the documentation

thanks again!

aduh95 commented on doc/api/assert.md:
````diff
@@ -2548,6 +2548,286 @@ assert.throws(throwingFirst, /Second$/);
+```mjs
+import assert from 'node:assert';
+
+assert.includes([1, 2, 3], ['2', 3]);
+// OK
````
As I was saying in the other comment, I don't really like this approach for arrays, IMO it should mirror the result of `assert.ok(actual.includes(expected))`, just like for strings.

```suggestion
assert.includes([1, 2, 3], ['2', 3]);
// AssertionError

assert.includes([1, 2, 3], '2');
// AssertionError

assert.includes([1, 2, 3], 2);
// OK
```

aduh95 replied:
It looks like the documentation still needs to be updated.

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +357,191 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+    throw new ERR_MISSING_ARGS('actual', 'expected');
+  }
+
+  if (typeof actual !== 'string' && !ArrayIsArray(actual)) {
+    throw new ERR_INVALID_ARG_TYPE('actual', ['string', 'Array'], actual);
+  }
+
+  if (!actual.includes(expected)) {
```
```suggestion
  const isString = typeof actual === 'string';

  if (!isString && !ArrayIsArray(actual)) {
    throw new ERR_INVALID_ARG_TYPE('actual', ['string', 'Array'], actual);
  }

  if (!(isString ? StringPrototypeIncludes : ArrayPrototypeIncludes)(actual, expected)) {
```

Can we add a test that includes a null-prototype array (`Object.setPrototypeOf([], null)`) to ensure `assert.includes` still works in this case?

aduh95 commented on doc/api/assert.md:
````diff
@@ -2548,6 +2548,226 @@ assert.throws(throwingFirst, /Second$/);
+```mjs
+import assert from 'node:assert';
+
+assert.includes([1, 2, 3], 2);
+// OK
````
We might as well add a example specifically about "non-strict" behavior not being there
```suggestion
assert.includes([1, 2, 3], 2);
// OK

assert.includes([1, 2, 3], '2');
// AssertionError
```

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +358,193 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+      return StringPrototypeIncludes(actual, expected);
+    }
+
+    return ArrayPrototypeEvery(expected, (item) => ArrayPrototypeSome(actual, (actualItem) => {
+      // eslint-disable-next-line eqeqeq
+      return loose ? actualItem == item : actualItem === item;
+    }));
```
```suggestion
    return ArrayPrototypeEvery(expected, (item) => loose ? ArrayPrototypeSome(actual, (actualItem) => {
      // eslint-disable-next-line eqeqeq
      return actualItem == item;
    }) : ArrayPrototypeIncludes(actual, item));
```

RedYetiDev commented:
Can you squash into a single commit with an updated message reflecting the current state?

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +358,197 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+
+  // Check for Set object equality
+  if (isSet(actual) && isSet(expected)) {
+    if (actual.size !== expected.size) {
```
The equality seems weird, wouldn't we want to support `actual` with more elements than `expected`?
```suggestion
    // TODO(aduh95): switch to `SetPrototypeIsSubsetOf` when it's available
    if (actual.size > expected.size) {
```
BTW (and sorry it's very late in the conversation, I didn't think of it earlier), `assert.isSubsetOf` might have been a better name to reflect to the built-in `Set` method (https://tc39.es/ecma262/#sec-set.prototype.issubsetof)

puskin94 replied:
Good catch, it was reminiscence from OP which I didn't notice

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +358,197 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+    return ArrayPrototypeEvery(expected, (item) => ArrayPrototypeIncludes(actual, item));
+  }
+
+  // Check non object types equality
+  if (!isPlainObject(actual) || !isPlainObject(expected)) {
```
I don't understand what "non object types" mean, could you elaborate? I find it a bit weird that we'd need a `isPlainObject` when (AFAIK) we have needed it before, I'm not sure it would be a good precedent – it seems to me all objects should be treated the same.

puskin94 replied:
This is a comment left from the OP of the MR that , tbh, confused me a bit as well when I first read it.

The block is supposed to execute whenever you are comparing types where `actual` AND `expected` are not objects

aduh95 replied:
If we're dealing with non-objects, the check should be `typeof … !== 'object'`, not that `isPlainObject` object (which is not documented and would be very hard to document)

puskin94 replied:
@aduh95 it is a lot more complicated than that...

types that would be considered objects:

1. regexes
2. Dates
3. Errors
4. nulls

and the list is much longer than I can possibly remember 😄 that function is just checking that it is going in the if block if one of those cases happens

I could separate all those cases, for readability, but it would be just for that

aduh95 replied:
It's not just readability, it's also about being consistent. `isPlainObject` is an anti-pattern IMO, if we want to e.g. special case regexes, let's do that explicitly.

puskin94 replied:
but we don't want to add special cases for e.g. regexes, we want the opposite, we only want to treat differently the case where you don't have 2 plain objects to compare, which is what the if statement is doing

aduh95 replied:
I'd argue there's no way to detect when an object is a plain object vs a special object – except on the case where it's a specific "other kind" of object (the ones that have a specific internal slot defined in the spec, e.g. arrays, regexes, proxies, etc.) – and so we should not try to define arbitrary conditions for what constitutes a plain object.

puskin94 replied:
I am getting a little confused then, what is the implementation you are suggesting?

RedYetiDev commented:
@puskin94 can you rebase to include #55036? You can undo your changes to `test-runner-assert`, as that PR should cover it.

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +358,198 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  }
+
+  // Comparison done when at least one of the values is not an object
+  if (!isPlainObject(actual) || !isPlainObject(expected)) {
```
```suggestion
  if (isSpecial(actual) || isSpecial(expected)) {
```

aduh95 commented on test/parallel/test-assert-objects.js:
```diff
@@ -0,0 +1,605 @@
+          actual: new Set(['value1', 'value2']),
+          expected: new Set(['value1', 'value3']),
+        },
+        {
+          description:
+            'throws when comparing plain objects from different realms',
+          actual: vm.runInNewContext(`({
+            a: 1,
+            b: 2n,
+            c: "3",
+            d: /4/,
+            e: new Set([5]),
+            f: [6],
+            g: new Uint8Array()
+          })`),
+          expected: { b: 2n, e: new Set([5]), f: [6], g: new Uint8Array() },
+        },
```
We definitely NOT want to throw in this case!

RedYetiDev commented on test/parallel/test-assert-objects.js:
```diff
@@ -0,0 +1,605 @@
+    describe('throws an error', () => {
+      [
+        {
+          description: 'throws because actual is nor an Array or a string',
```
```suggestion
          description: 'throws because actual is not an Array nor a string',
```

RedYetiDev commented on test/parallel/test-assert-objects.js:
```diff
@@ -0,0 +1,605 @@
+        },
+        {
+          description:
+            'throws because using includes with a string and an array',
```
```suggestion
            'throws because an string is being compared to an array',
```

RedYetiDev commented on test/parallel/test-assert-objects.js:
```diff
@@ -0,0 +1,605 @@
+  });
+
+  describe('includes', () => {
+    describe('throws an error', () => {
```
Can you add the following cases:
1. `expected` is not a valid value
2. `actual` array is being compared to `expected` string

puskin94 replied:
1. `expected` is `any`, added a test for when `expected` is missing
2: 👍

puskin94 commented on lib/assert.js:
```diff
@@ -341,6 +358,198 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
   }
 };
 
+function isPlainObject(obj) {
+  if (typeof obj !== 'object' || obj === null) {
+    return false;
+  }
+  const proto = ObjectGetPrototypeOf(obj);
+  return proto === ObjectPrototype || proto === null || ObjectPrototypeToString(obj) === '[object Object]';
+}
```
ah I get it now!  
so you prefer to outline every single case where it is NOT an `object` instead of when it IS an `object`, I am not sure I like this approach tbh 😃

puskin94 replied:
@aduh95 ahhh now I see what you mean 😃 sorry for the confusion. I thought the list of cases would have gone longer, reading the proposal unreadable 😄

aduh95 commented on test/parallel/test-assert-objects.js:
```diff
@@ -0,0 +1,617 @@
+      },
+      {
+        description:
+          'throws when comparing plain objects from different realms',
```
```suggestion
          'compares plain objects from different realms',
```

aduh95 commented on test/parallel/test-assert-objects.js:
```diff
@@ -0,0 +1,617 @@
+        },
+        {
+          description:
+            'throws because an string is being compared to an array',
```
```suggestion
            'throws because a string is being compared to an array',
```

rluvaton commented on lib/assert.js:
```diff
@@ -341,6 +355,194 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  const keysExpected = ReflectOwnKeys(expected);
+
+  // Handle circular references
+  if (SetPrototypeHas(comparedObjects, actual)) {
```
Should we instead use WeakSet instead for better memory?

puskin94 replied:
@rluvaton I didn't know about the difference in methods! updated, thanks 😄

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +357,194 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  const keysExpected = ReflectOwnKeys(expected);
+
+  // Handle circular references
+  if (WeakSetPrototypeHas(comparedObjects, actual)) {
```
```suggestion
  if (comparedObjects.has(actual)) {
```

puskin94 replied:
@aduh95 how come this time not a primordial?

aduh95 replied:
https://github.com/nodejs/node/blob/c42d8461b061136da2b3c04bd976d2e15b81969d/doc/contributing/primordials.md#L94-L99

There's no reason to use primordials on a safe class, they're designed to be protected from prototype pollution.

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +357,194 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  }
+
+  // Check if expected array is a subset of actual array
+  if (ArrayIsArray(actual) && ArrayIsArray(expected)) {
+    if (expected.length > actual.length) {
+      return false;
+    }
+    if (typeof actual === 'string') {
+      return StringPrototypeIncludes(actual, expected);
+    }
+
```
I doubt that `Array.isArray(actual)` and `typeof actual === 'string'` can be true at the same time – and even if that's possible, I don't think we want to support that use-case.
```suggestion
  if (ArrayIsArray(actual) && ArrayIsArray(expected)) {
    if (expected.length > actual.length) {
      return false;
    }

```

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +357,194 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  actual,
+  expected,
+  loose = false,
+  comparedObjects = new SafeWeakSet(),
```
nit: I think we could create the `WeakSet` only in the branch that needs it, which should speed up the other branches

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +356,199 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  // Use Reflect.ownKeys() instead of Object.keys() to include symbol properties
+  const keysExpected = ReflectOwnKeys(expected);
+
+  if (!comparedObjects) {
+    comparedObjects = new SafeWeakSet();
+  }
```
```suggestion
  comparedObjects ??= new SafeWeakSet();
```

boukeversteegh commented:
Nice, looking forward to this getting merged!

puskin94 commented:
@atlowChemi uuh I didn't know it! thanks for showing me!

updated the code to keep `common.hasCrypto` in consideration

BridgeAR reviewed (changes requested):
I just look at this now and I could not find any comments about why we want to have two methods `includes()` and `partialDeepEqual()`. I think it would be nicer to simplify that and just have `partialDeepEqual()` as it should IMO just do the same what `includes()` currently does.

Could someone please point me to some discussion around that?

RedYetiDev commented:
Also, `includes()` does not work on buffers, should it?

RedYetiDev commented on doc/api/assert.md:
````diff
@@ -2548,6 +2548,232 @@ assert.throws(throwingFirst, /Second$/);
+// AssertionError
+```
+
+## `assert.includes(actual, expected[, message])`
````
WDYT about `includes` and `includesStrict`?

`includesStrict([obj, otherObj], obj)` -> True
`includesStrict([{}], {})` -> False

But

`includes([obj, otherObj], obj)` -> True
`includes([{}], {})` -> True

aduh95 replied:
See https://github.com/nodejs/node/pull/54630#discussion_r1765880973

BridgeAR replied:
The name does not tell anything about how the comparison works. The original one was about strict equality as in three equal signs. This is about reference equality and I don't think we should do this.

BridgeAR commented on doc/api/assert.md:
```diff
@@ -2548,6 +2548,232 @@ assert.throws(throwingFirst, /Second$/);
 Due to the confusing error-prone notation, avoid a string as the second
 argument.
 
+## `assert.partialDeepEqual(actual, expected[, message])`
```
The existence of the `strict` method is something that is very confusing and I strongly recommend we do not add any "non"-strict version. I could not find anyone asking for this either.

That should never be used. If we ever want to have a different comparison, it should be done by e.g., `assert.any(Number)` or similar.

jsumners replied:
I give you a whole discussion around wanting this https://github.com/fastify/fastify/discussions/5628

BridgeAR replied:
@jsumners I am absolutely +1 on adding some kind of partial deep equal comparison as done here. This PR just adds three different APIs while I believe we should only ship the `partialDeepStrictEqual` one. I am not a super fan of the name but I know there was a lot of back and forth and it is tricky to find the right name.

@puskin94 what about doing the following as next steps:

1. Remove everything besides the current `partialDeepStrictEqual()` implementation.
2. Have a small poll for the name.

That way we could ship the main requested feature and later on decide if we want to add anything else on top.

I think this is great work and we should definitely ship that functionality!

puskin94 commented:
Small poll about the name of the `partialDeepStrictEqual` method, which allows to perform subset comparisons as:

`assert.partialDeepStrictEqual({ a: 1, b: 2, c: 3 }, { a: 1, b: 2 });`

1️⃣  partialDeepStrictEqual  
2️⃣  subsetDeepStrictEqual  
3️⃣  matchStrictObject  
4️⃣  deepStrictMatch  
5️⃣  ??? (please write the new proposed name)

You can use the number emojis on this comment to vote your favorite!

puskin94 commented:
Small poll about the name of the `partialDeepStrictEqual` method, which allows to perform subset comparisons as:

`assert.partialDeepStrictEqual({ a: 1, b: 2, c: 3 }, { a: 1, b: 2 });`

👍   partialDeepStrictEqual  
😄   subsetDeepStrictEqual  
🎉   matchStrictObject  
❤️   deepStrictMatch  
🚀   ??? (please write the new proposed name)

You can use the emojis above to react on this comment to vote your favorite!

boukeversteegh commented:
> Small poll about the name of the `partialDeepStrictEqual` method, which allows to perform subset comparisons as:
> 
> `assert.partialDeepStrictEqual({ a: 1, b: 2, c: 3 }, { a: 1, b: 2 });`
> 
> 👍 partialDeepStrictEqual
> 😄 subsetDeepStrictEqual
> 🎉 matchStrictObject
> ❤️ deepStrictMatch
> 🚀 ??? (please write the new proposed name)
> 
> You can use the emojis above to react on this comment to vote your favorite!

To determine the best name, I'm thinking of the structure of the method name by looking at two main components:

- **The Operator**: What is the primary action or assertion being performed? Examples include "equals" or "match".
- **Modifiers**: What options modify the evaluation? Examples include "deep" or "strict".

It's conventional to place the operator at the end of the method name (e.g., deepEqual). The key question is whether this method is a modification of the existing "Equal" operator or if it represents a new operator altogether.

If it's a modification of "Equal," the method name would logically be constructed as:

- `[Modifiers]` + `[Partial]` + `Equal`

Regarding the modifiers:

- **Depth**: Since the assertion inherently recurses through nested objects, "deep" might be redundant.
- **Strictness**: "Strict" is relevant because it specifies how the property values are compared (using strict equality).

With this in mind, a possible name could be `strictPartialEqual`, aligning with existing naming conventions like `strictEqual`.

However, given that this method operates differently from a standard equality check—specifically, it checks for a subset rather than full equality—it might be better to consider a new operator. The goal is to find an operator that implies recursive comparison and ideally is a single word.

Possible operators and considerations:

- **Match**: Could be confused with regex matching.
- **MatchObject**: More explicit and reduces confusion.
- **Subset**: Meaning commonly understood, although set-subsets and recursive property subsets are quite different.
- **Is Subset**: Since verbs are typically omitted in assertion method names, not the best choice.
- **Subset Equal**: "Equal" implies a different operation so not the best choice either.
- **Includes**: Might not clearly convey recursive comparison, as it's commonly associated with arrays or strings.
- **Contains**: Not commonly used in JavaScript for this purpose.
- **Has**: Lacks clarity about the depth and nature of the assertion

Final suggestions:

- `strictMatchObject` -- Match Object implies recursion, so "deep' is not needed
- `deepStrictSubset` -- Adding deep to allow for non recursive subsets like for lists
- `deepStrictIncludes` -- Adding deep to clarify the recursive nature of the assertion

danielbayley commented:
> Small poll about the name of the `partialDeepStrictEqual` method, which allows to perform subset comparisons

@puskin94 See my reasoning for `deep`[`Strict`]`Match` in previous https://github.com/nodejs/node/issues/50399#issuecomment-2296531960:

> > `assert.matchObject`/[`matchObjectStrict`](https://github.com/nodejs/node/pull/53415#top)
> 
> 
> 
> This way, if we also wanted a corresponding method to match array subsets, it would have to be called `assert.matchArray` or something…
> 
> 
> 
> Given that we currently have `deepEqual` for comparing objects _and_ arrays for exact equality, `match` for matching substrings, and considering existing parlance… I think we should go with the name `deepMatch`, to assert that A (object _or_ array) contains subset B. We could/should also have an `assert.includes` method that mirrors the behaviour of the native JS `.includes` (string or array).
> 
> 
> 
> > `assert.contains`
> 
> 
> 
> ‘contains’ makes sense, but the naming should match existing JavaScript parlance, which, for reasons, went with `.includes` instead. Aiming for consistency here…
> 
> 
> 
> To clarify, something like:
> 
> ~~~ js
> 
> import assert from "node:assert/strict"
> 
> 
> 
> assert.deepMatch({ a: 1, b: 2, c: 3 }, { b: 2 }) // pass
> 
> assert.deepMatch([1, 2, 3], [2]) // pass
> 
> ~~~
> 
> 
> 
> Then, for `assert.includes`, something like:
> 
> ~~~ js
> 
> import assert from "node:assert/strict"
> 
> 
> 
> assert.includes ??= (a, b) => assert.equal(a.includes(b), true)
> 
> 
> 
> assert.includes("abc", "b")   // pass
> 
> assert.includes([1, 2, 3], 2) // pass
> 
> ~~~
> 
> 
> 
> @arthurfiorette @targos @ljharb @MoLow @reconbot @synapse @simoneb What do you guys think?

BridgeAR commented:
Using `strict` is IMO not needed. The original usage was already not ideal and it is still confusing people.
I think just dropping that overall would be best.

puskin94 commented:
> Using `strict` in the name is IMO not needed. The original usage was already not ideal and it is still confusing people. I think just dropping that overall would be best.

I agree with you, I would remove it too, but wouldn't it be more confusing to remove the `strict` part here and keeping it in the other methods? If you do that, I would assume this is loose

aduh95 commented:
If you don't want `strict` in the name, you can import it from `node:assert/strict`. Removing it from the exported name in `node:assert` would be confusing IMO.

puskin94 commented:
How do we want to move with this one?

pmarchini commented:
> Small poll about the name of the `partialDeepStrictEqual` method, which allows to perform subset comparisons as:
> 
> `assert.partialDeepStrictEqual({ a: 1, b: 2, c: 3 }, { a: 1, b: 2 });`
> 
> 👍 partialDeepStrictEqual 😄 subsetDeepStrictEqual 🎉 matchStrictObject ❤️ deepStrictMatch 🚀 ??? (please write the new proposed name)
> 
> You can use the emojis above to react on this comment to vote your favorite!

I've seen this PR open for quite a while. Before this PR, I noticed a very similar discussion in another PR, from which this one originated (https://github.com/nodejs/node/pull/53415).

I think that this feature would add significant value to the assertions set, so I strongly believe we should unblock this by deciding on a name.  
Considering that this discussion has already been going on for quite some time, I think it’s time to ask for a TSC vote.

@nodejs/tsc

aduh95 commented:
@pmarchini are you happy with the proposal in https://github.com/nodejs/node/pull/54630#issuecomment-2410311067, or is there a more up-to-date list if it does go to a TSC vote?

> Final suggestions:
> 
> * `strictMatchObject` -- Match Object implies recursion, so "deep' is not needed
> * `deepStrictMatch` -- As suggested, using deep to clarify recursive nature and distinguish it from regex
> * `deepStrictSubset` -- Adding deep to allow for non recursive subsets like for lists
> * `deepStrictIncludes` -- Adding deep to clarify the recursive nature of the assertion
> * `strictPartialEqual` -- Only if the assertion is considered a variation of equal, which it isn't, strictly speaking

pmarchini commented:
Hey @aduh95, absolutely, I also agree with considering partialDeepStrictEqual! 😁  
I don’t think there’s a more recent list of proposals than the one you just quoted.  
(I just realised I quoted the wrong comment)

IMHO I think we should go with that one, but I’d suggest that everyone involved in this discussion be able to give a "final word" on it by a set deadline 🚀

puskin94 commented on lib/assert.js:
```diff
@@ -341,6 +355,143 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+function compareBranch(
+  actual,
+  expected,
+  loose = false,
```
@BridgeAR good catch! thanks for the additional review!

boukeversteegh commented:
I would consider two more candidates for the voting list.

- All previous suggestions without STRICT in the name, assuming we can solve this with a separate import node strict as mentioned
- `match` - simply one single term. Although it's not consistent with the rest of the codebase, I do feel like this is going to be one of the most popular assertions in this library and a simple name invites developers to use it

BridgeAR commented:
@boukeversteegh
> I would consider two more candidates for the voting list.
> All previous suggestions without STRICT in the name, assuming we can solve this with a separate import node strict as mentioned
match - simply one single term. Although it's not consistent with the rest of the codebase, I do feel like this is going to be one of the most popular assertions in this library and a simple name invites developers to use it

We could add the name to the strict export only. That would keep it consistent with the other names.
I would just do that as a separate contribution to get this through the door.

BridgeAR commented on lib/assert.js:
```diff
@@ -341,6 +354,130 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+      return false;
+    }
+
+    return ArrayPrototypeEvery(expected, (item) => ArrayPrototypeIncludes(actual, item));
```
This is actually incorrect. It will match identical parts multiple times. It is also going to have a bad performance for these cases.

I am not going to block this though, to get the general idea in and to improve afterwards. Can we just mark it experimental for now, due to these issues?

BridgeAR commented on lib/assert.js:
```diff
@@ -341,6 +354,130 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+    const key = keysExpected[i];
+    assert(
+      ReflectHas(actual, key),
+      new AssertionError({ message: `Expected key ${String(key)} not found in actual object` }),
```
This message won't be ideal. If it's a deeper array, it is completely unclear what this belongs to. It was actually my main issue with my first attempt. To highlight what parts are missing is tricky.

BridgeAR commented on lib/assert.js:
```diff
@@ -341,6 +354,130 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  if (comparedObjects.has(actual)) {
+    return true;
+  }
+  comparedObjects.add(actual);
```
I did not verify that but I believe the circular structure won't always work this way. AFAIC we have to handle both sides similar to the current implementation.

tniessen reviewed (commented):
I think the documentation should precisely characterize the behavior of this function, including edge cases. Right now, it only mentions the "main difference", but that doesn't explain differences such as the following:

```js
assert.deepStrictEqual(new Set([{ a: 1 }]), new Set([{ a: 1 }]))
// No error.

assert.partialDeepStrictEqual(new Set([{ a: 1 }]), new Set([{ a: 1 }]))
// Uncaught AssertionError.
```

On a side note, the commit message should use an imperative verb (e.g., `add`) instead of `adds` to comply with our guidelines.

BridgeAR commented:
@tniessen I consider the mentioned difference a bug in the current implementation. I see a couple of cases not yet working as expected. We can either make sure it works fine up front or land this and continue on it later.

puskin94 commented:
Fixed the issue with the comparison between objects in Sets, updated the commit message and the documentation

tniessen commented:
> Fixed the issue with the comparison between objects in Sets

I checked out f0722f9cbbaee5c2e8d8383acab7c84043788339. Is this expected behavior?

```js
assert.partialDeepStrictEqual(new Set([{ a: 1 }]), new Set([{ a: 1 }, { b: 1 }]))
// No error.
```

puskin94 commented:
> assert.partialDeepStrictEqual(new Set([{ a: 1 }]), new Set([{ a: 1 }, { b: 1 }]))

you are right, I might have introduced that issue in the last commit. Fixed and added relevant tests :)

marco-ippolito reviewed (approved):
LGTM

tniessen reviewed (commented):
I still think this function deserves a precise definition of its behavior. The two `Set`-related bugs that I discovered today could have been expected behavior for all I know.

Looking at the implementation, objects in `Set`s are apparently not tested for reference-equality, but objects in `Array`s are. Again, I have no idea if that is expected behavior or not.

```js
assert.deepStrictEqual([{ a: 5 }], [{ a: 5 }])
// No error.

assert.partialDeepStrictEqual([{ a: 5 }], [{ a: 5 }])
// AssertionError.
```

Also, the order of elements of `Array`s is ignored for the purpose of this assertion, and holes in sparse arrays are ignored as well. I again have no idea if that is expected or not, but as a user, I certainly wouldn't expect it based on the documentation.

puskin94 commented:
> I still think this function deserves a precise definition of its behavior. The two `Set`-related bugs that I discovered today could have been expected behavior for all I know.
> 
> Looking at the implementation, objects in `Set`s are apparently not tested for reference-equality, but objects in `Array`s are. Again, I have no idea if that is expected behavior or not.
> 
> ```js
> assert.deepStrictEqual([{ a: 5 }], [{ a: 5 }])
> // No error.
> 
> assert.partialDeepStrictEqual([{ a: 5 }], [{ a: 5 }])
> // AssertionError.
> ```
> 
> Also, the order of elements of `Array`s is ignored for the purpose of this assertion, and holes in sparse arrays are ignored as well. I again have no idea if that is expected or not, but as a user, I certainly wouldn't expect it based on the documentation.

I would consider what you found an issue, I will fix the code soon, but I would like other people to chip in on how this case should work according to them

BridgeAR commented:
I think this could land as experimental and we explicitly mention that this should always pass `deepStrictEqual()`. So it's like a superset of that. I also think we should verify that later on and improving the code accordingly.

@puskin94 would you mind adding that?

tniessen commented:
> I think this could land as experimental and we explicitly mention that this should always pass `deepStrictEqual()`. So it's like a superset of that.

I agree with the general strategy, but I think it'd be a good idea for the documentation to accurately define what superset of input pairs are permitted that are not allowed by `deepStrictEqual`. Otherwise, users have no way of knowing whether they are relying on expected behavior or on bugs.

BridgeAR commented:
I think it will still be refined during the development. And while doing that, we should define it.
It is marked as experimental because it's not yet where it should be.

danielbayley commented:
> > I still think this function deserves a precise definition of its behavior. The two `Set`-related bugs that I discovered today could have been expected behavior for all I know.
> > Looking at the implementation, objects in `Set`s are apparently not tested for reference-equality, but objects in `Array`s are. Again, I have no idea if that is expected behavior or not.
> > ```js
> > assert.deepStrictEqual([{ a: 5 }], [{ a: 5 }])
> > // No error.
> > 
> > assert.partialDeepStrictEqual([{ a: 5 }], [{ a: 5 }])
> > // AssertionError.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also, the order of elements of `Array`s is ignored for the purpose of this assertion, and holes in sparse arrays are ignored as well. I again have no idea if that is expected or not, but as a user, I certainly wouldn't expect it based on the documentation.
> 
> I would consider what you found an issue, I will fix the code soon, but I would like other people to chip in on how this case should work according to them

Would these different behaviours not be best expressed as boolean props of an optional object passed as the last argument to these functions? Something like:
~~~ js
assert.deepStrictEqual([1, 2, 3], [3, 1, 2]) // pass

assert.deepStrictEqual([1, 2, 3], [3, 1, 2], { order: true }) // fail
~~~

tniessen commented:
@danielbayley The problem with such options is that it's unclear whether or not they should apply to nested decision procedures. Neither option will cover all use cases.

aduh95 commented:
Tests are failing

BridgeAR commented:
Since we have no one blocking, I would go ahead and land this as is in the next few days.
That way we are able to iterate on the implementation details from that point on. If someone fears the current implementation and wants to have a stronger warning besides the experimental part: we could mention that the algorithm is currently worked upon and will still change.

marco-ippolito reviewed (changes requested):
This needs an experimental warning

BridgeAR commented:
Adding the warning sounds good.
The experimental status is somewhat in-between 1.0 and 1.1. I suggest to keep it as 1.0 to make sure breaking changes are reflected best.

marco-ippolito commented on lib/assert.js:
```diff
@@ -341,6 +364,191 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  expected,
+  message,
+) {
+  emitExperimentalWarning('partialDeepStrictEqual');
```
```suggestion
  emitExperimentalWarning('assert.partialDeepStrictEqual');
```

aduh95 commented on lib/assert.js:
```diff
@@ -341,6 +366,190 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
+  }
+
+  // Check for Set object equality
+  // TODO(aduh95): switch to `SetPrototypeIsSubsetOf` when it's available
```
It was added back when `assert.includes` was part of this PR. I'll make a follow up PR

BridgeAR commented:
@puskin94 thank you for sticking to this and to continuesly improve everything so that it was able to land!
Great contribution!

puskin94 commented:
> @puskin94 thank you for sticking to this and to continuesly improve everything so that it was able to land! Great contribution!

thank you too @BridgeAR for the huge help you were able to provide me! It would have been several order of magnitude more difficult to land this without your support 🥲

hamzabhatti5253 commented on doc/api/assert.md:
```diff
@@ -2548,6 +2548,81 @@ assert.throws(throwingFirst, /Second$/);
 Due to the confusing error-prone notation, avoid a string as the second
 argument.
 
+## `assert.partialDeepStrictEqual(actual, expected[, message])`
```
Apply

hamzabhatti5253 replied:
Apply

hamzabhatti5253 replied:
Apply
